### ISSUE TITLE
catkin_make error with zsh shell

### ISSUE THREAD

**Issue by prfraanje** [ros/catkin#652]
when doing a catkin_make on a brand new initialized catkin workspace in a zsh environment I get the error:

```
....
-- Found gtest sources under '/usr/src/gtest': gtests will be built
-- catkin 0.5.86
/home/rufus/catkin/devel/env.sh: 6: /opt/ros/hydro/share/rosbash/roszsh: function: not found
/home/rufus/catkin/devel/env.sh: 7: /opt/ros/hydro/share/rosbash/roszsh: [[: not found
/home/rufus/catkin/devel/env.sh: 7: /opt/ros/hydro/share/rosbash/roszsh: Linux: not found
sed: invalid option -- 'c'
Usage: sed [OPTION]... {{script-only-if-no-other-script}} [input-file]...
.....
GNU sed home page: <http://www.gnu.org/software/sed/>.
General help using GNU software: <http://www.gnu.org/gethelp/>.
/home/rufus/catkin/devel/env.sh: 12: /opt/ros/hydro/share/rosbash/roszsh: Syntax error: "}}" unexpected
Traceback (most recent call last):
  File "/home/rufus/catkin/build/catkin_generated/generate_cached_setup.py", line 21, in <module>
code = generate_environment_script('/home/rufus/catkin/devel/env.sh')
  File "/opt/ros/hydro/lib/python2.7/dist-packages/catkin/environment_cache.py", line 61, in generate_environment_script
output = subprocess.check_output([env_script, sys.executable, '-c', python_code])
  File "/usr/lib/python2.7/subprocess.py", line 544, in check_output
raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['/home/rufus/catkin/devel/env.sh', '/usr/bin/python', '-c', 'import os; print(dict(os.environ))']' returned non-zero exit status 2
CMake Error at /opt/ros/hydro/share/catkin/cmake/safe_execute_process.cmake:11 (message):
  execute_process(/usr/bin/python
  "/home/rufus/catkin/build/catkin_generated/generate_cached_setup.py")
  returned error code 1
Call Stack (most recent call first):
  /opt/ros/hydro/share/catkin/cmake/all.cmake:178 (safe_execute_process)
  /opt/ros/hydro/share/catkin/cmake/catkinConfig.cmake:20 (include)
  CMakeLists.txt:52 (find_package)
-- Configuring incomplete, errors occurred!
make: *** [cmake_check_build_system] Error 1
Invoking "make cmake_check_build_system" failed
```

I think something goes wrong in generating the right zsh environment. I analyzed a bit further by evaluating lines 11 and 12 in devel/env.sh and got similar errors. Then by changing setup.zsh on line 12 in devel/env.sh, I got errors like builtin: not found and emulate: not found. Finally changing line 1 in devel/env.sh  into #!/usr/bin/env zsh seems to work properly. So changing catkin/cmake/templates/env.sh.in accordingly helped me to run catkin_make properly. But my changes are rather ad hoc, and I'm not sure everything is fixed all right (e.g. still got an issue on setup_cached.zsh: /home/rufus/catkin/build/catkin_generated/setup_cached.zsh:.:12: no such file or directory: /home/rufus/catkin/build/catkin_generated/setup_cached.zsh). I would be grateful for any fixes / or suggestions on things I need to change. Thanks!

**Comment by dirk-thomas** [comment]
Please provide more information to reproduce the problem. Are you using Ubuntu? Do you have ROS installed from Debian packages? Have you sourced `/opt/ros/hydro/setup.zsh` before?

I tried the following sequence (with Indigo though) and it worked fine for me:

```
zsh
. /opt/ros/indigo/setup.zsh
mkdir -p /tmp/test_zsh/src
cd /tmp/test_zsh/src
catkin_create_pkg foo
cd /tmp/test_zsh
catkin_make
. devel/setup.zsh
```

Can you please try the same sequence and post you output?

**Comment by wjwwood** [comment]
I can also not reproduce this on OS X with zsh.

**Comment by prfraanje** [comment]
Thank you for addressing my issue.
Evaluating the commands I still get the same errors:

```
vm-ros:~$ zsh
vm-ros:~$ . /opt/ros/hydro/setup.zsh
vm-ros:~$ mkdir -p /tmp/test_zsh/src
vm-ros:~$ cd /tmp/test_zsh/src
vm-ros:src$ catkin_create_pkg foo
Created file foo/package.xml
Created file foo/CMakeLists.txt
Successfully created files in /tmp/test_zsh/src/foo. Please adjust the values in package.xml.
vm-ros:src$ cd /tmp/test_zsh
vm-ros:test_zsh$ catkin_make
Base path: /tmp/test_zsh
Source space: /tmp/test_zsh/src
Build space: /tmp/test_zsh/build
Devel space: /tmp/test_zsh/devel
Install space: /tmp/test_zsh/install
Creating symlink "/tmp/test_zsh/src/CMakeLists.txt" pointing to "/opt/ros/hydro/share/catkin/cmake/toplevel.cmake"
####
#### Running command: "cmake /tmp/test_zsh/src -DCATKIN_DEVEL_PREFIX=/tmp/test_zsh/devel -DCMAKE_INSTALL_PREFIX=/tmp/test_zsh/install" in "/tmp/test_zsh/build"
####
-- The C compiler identification is GNU
-- The CXX compiler identification is GNU
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Using CATKIN_DEVEL_PREFIX: /tmp/test_zsh/devel
-- Using CMAKE_PREFIX_PATH: /opt/ros/hydro
-- This workspace overlays: /opt/ros/hydro
-- Found PythonInterp: /usr/bin/python (found version "2.7.3")
-- Using PYTHON_EXECUTABLE: /usr/bin/python
-- Python version: 2.7
-- Using Debian Python package layout
-- Using CATKIN_ENABLE_TESTING: ON
-- Call enable_testing()
-- Using CATKIN_TEST_RESULTS_DIR: /tmp/test_zsh/build/test_results
-- Looking for include files CMAKE_HAVE_PTHREAD_H
-- Looking for include files CMAKE_HAVE_PTHREAD_H - found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE 
-- Found gtest sources under '/usr/src/gtest': gtests will be built
-- catkin 0.5.86
/tmp/test_zsh/devel/env.sh: 6: /opt/ros/hydro/share/rosbash/roszsh: function: not found
/tmp/test_zsh/devel/env.sh: 7: /opt/ros/hydro/share/rosbash/roszsh: [[: not found
/tmp/test_zsh/devel/env.sh: 7: /opt/ros/hydro/share/rosbash/roszsh: Linux: not found
sed: invalid option -- 'c'
Usage: sed [OPTION]... {{script-only-if-no-other-script}} [input-file]...

  -n, --quiet, --silent
                 suppress automatic printing of pattern space
  -e script, --expression=script
                 add the script to the commands to be executed
  -f script-file, --file=script-file
                 add the contents of script-file to the commands to be executed
  --follow-symlinks
                 follow symlinks when processing in place
  -i[SUFFIX], --in-place[=SUFFIX]
                 edit files in place (makes backup if extension supplied)
  -l N, --line-length=N
                 specify the desired line-wrap length for the `l' command
  --posix
                 disable all GNU extensions.
  -r, --regexp-extended
                 use extended regular expressions in the script.
  -s, --separate
                 consider files as separate rather than as a single continuous
                 long stream.
  -u, --unbuffered
                 load minimal amounts of data from the input files and flush
                 the output buffers more often
      --help     display this help and exit
      --version  output version information and exit

If no -e, --expression, -f, or --file option is given, then the first
non-option argument is taken as the sed script to interpret.  All
remaining arguments are names of input files; if no input files are
specified, then the standard input is read.

GNU sed home page: <http://www.gnu.org/software/sed/>.
General help using GNU software: <http://www.gnu.org/gethelp/>.
/tmp/test_zsh/devel/env.sh: 12: /opt/ros/hydro/share/rosbash/roszsh: Syntax error: "}}" unexpected
Traceback (most recent call last):
  File "/tmp/test_zsh/build/catkin_generated/generate_cached_setup.py", line 21, in <module>
    code = generate_environment_script('/tmp/test_zsh/devel/env.sh')
  File "/opt/ros/hydro/lib/python2.7/dist-packages/catkin/environment_cache.py", line 61, in generate_environment_script
    output = subprocess.check_output([env_script, sys.executable, '-c', python_code])
  File "/usr/lib/python2.7/subprocess.py", line 544, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['/tmp/test_zsh/devel/env.sh', '/usr/bin/python', '-c', 'import os; print(dict(os.environ))']' returned non-zero exit status 2
CMake Error at /opt/ros/hydro/share/catkin/cmake/safe_execute_process.cmake:11 (message):
  execute_process(/usr/bin/python
  "/tmp/test_zsh/build/catkin_generated/generate_cached_setup.py") returned
  error code 1
Call Stack (most recent call first):
  /opt/ros/hydro/share/catkin/cmake/all.cmake:178 (safe_execute_process)
  /opt/ros/hydro/share/catkin/cmake/catkinConfig.cmake:20 (include)
  CMakeLists.txt:52 (find_package)


-- Configuring incomplete, errors occurred!
Invoking "cmake" failed
vm-ros:test_zsh$ 
```

I've checked this with both bash and zsh as the default shell.
I'm running ROS Hydro in a Ubuntu 12.04 32bit virtual machine, I retrieved from http://nootrix.com/00download/download.html?fileId=rosHydroVM.

**Comment by jbohren** [comment]
@prfraanje Can you dump the output of `env` into a gist: https://gist.github.com/ ?

**Comment by prfraanje** [comment]
Yes, here you can find the output of env: https://gist.github.com/prfraanje/ed95eb32cbfa7b5f2e70

**Comment by jbohren** [comment]
As I suspected you're trying to source the zsh env files but it appears you're currently running bash: https://gist.github.com/prfraanje/ed95eb32cbfa7b5f2e70#file-env-L4

Before sourcing the `setup.zsh` file, you need to be running zsh.

**Comment by prfraanje** [comment]
I changed my default shell from zsh again into bash, that's why my SHELL=bash, the output of ps yields:

```
vm-ros:test_zsh$ ps
  PID TTY          TIME CMD
 2824 pts/2    00:00:00 bash
 2975 pts/2    00:00:00 zsh
 3225 pts/2    00:00:00 ps
```

Setting my default shell to zsh, logging out and in again, still gives me the same problem...

**Comment by jbohren** [comment]
> Setting my default shell to zsh, logging out and in again, still gives me the same problem...

And if you run `zsh` explicitly from the shell and then try it?

**Comment by prfraanje** [comment]
Thanks for the suggestion, but it doesn't help. I changed my default shell into zsh, indeed now we get:

```
vm-ros:test_zsh$ echo $SHELL
/bin/zsh
vm-ros:test_zsh$ ps
  PID TTY          TIME CMD
 3561 pts/2    00:00:00 zsh
 3659 pts/2    00:00:00 zsh
 3854 pts/2    00:00:00 ps
```

Note, that now I get two times zsh in the output of ps, because I'm running zsh a second time in the shell.

**Comment by jbohren** [comment]
> Thanks for the suggestion, but it doesn't help.

Each time you tried this, did you delete your `build` and `devel` directories?

**Comment by prfraanje** [comment]
sorry for my late reply, I was teaching all day. But yes, I performed the whole procedure as outlined by @dirk-thomas several times, after each time I perform a rm -rf /tmp/test_zsh

**Comment by prfraanje** [comment]
Could it be, that the ./devel/env.sh does try to evaluate in the sh shell rather than zsh? I've debugged a bit further (see below) and it seems that evaluating $. "/opt/ros/hydro/share/rosbash/roszsh" in the sh-shell doesn't work (in zsh-shell it does, as expected). However env.sh switches to sh, than evaluates ./devel/setup.sh still in sh-shell, which than tries to load "/opt/ros/hydro/share/rosbash/roszsh", which gives the error.
A solution could be for zsh to generate an ./devel/env.zsh rather than ./devel/env.sh, which then runs in zsh-shell and will evaluate e.g. ./devel/setup.zsh..... ?
(Note that the strange symbols %{%}%M:%1~%(!.#.$)%{%} in my prompt in sh-shell are because PS1 is not well interpreted by sh.)

```

-- Configuring incomplete, errors occurred!
Invoking "cmake" failed
vm-ros:test_zsh$ cd devel 
vm-ros:devel$ echo $CATKIN_SHELL
zsh
vm-ros:devel$ env sh
%{%}%M:%1~%(!.#.$)%{%} echo $CATKIN_SHELL
zsh
%{%}%M:%1~%(!.#.$)%{%} . /tmp/test_zsh/devel/setup.sh
sh: 6: /opt/ros/hydro/share/rosbash/roszsh: function: not found
sh: 7: /opt/ros/hydro/share/rosbash/roszsh: [[: not found
sh: 7: /opt/ros/hydro/share/rosbash/roszsh: Linux: not found
Usage: sed [OPTION]... {script-only-if-no-other-script} [input-file]...

  -n, --quiet, --silent
                 suppress automatic printing of pattern space
  -e script, --expression=script
                 add the script to the commands to be executed
  -f script-file, --file=script-file
                 add the contents of script-file to the commands to be executed
  --follow-symlinks
                 follow symlinks when processing in place
  -i[SUFFIX], --in-place[=SUFFIX]
                 edit files in place (makes backup if extension supplied)
  -l N, --line-length=N
                 specify the desired line-wrap length for the `l' command
  --posix
                 disable all GNU extensions.
  -r, --regexp-extended
                 use extended regular expressions in the script.
  -s, --separate
                 consider files as separate rather than as a single continuous
                 long stream.
  -u, --unbuffered
                 load minimal amounts of data from the input files and flush
                 the output buffers more often
      --help     display this help and exit
      --version  output version information and exit

If no -e, --expression, -f, or --file option is given, then the first
non-option argument is taken as the sed script to interpret.  All
remaining arguments are names of input files; if no input files are
specified, then the standard input is read.

GNU sed home page: <http://www.gnu.org/software/sed/>.
General help using GNU software: <http://www.gnu.org/gethelp/>.
sh: 12: /opt/ros/hydro/share/rosbash/roszsh: Syntax error: "}" unexpected
%{%}%M:%1~%(!.#.$)%{%} . "/opt/ros/hydro/share/rosbash/roszsh"
sh: 6: /opt/ros/hydro/share/rosbash/roszsh: function: not found
sh: 7: /opt/ros/hydro/share/rosbash/roszsh: [[: not found
sh: 7: /opt/ros/hydro/share/rosbash/roszsh: Linux: not found
Usage: sed [OPTION]... {script-only-if-no-other-script} [input-file]...

  -n, --quiet, --silent
                 suppress automatic printing of pattern space
  -e script, --expression=script
                 add the script to the commands to be executed
  -f script-file, --file=script-file
                 add the contents of script-file to the commands to be executed
  --follow-symlinks
                 follow symlinks when processing in place
  -i[SUFFIX], --in-place[=SUFFIX]
                 edit files in place (makes backup if extension supplied)
  -l N, --line-length=N
                 specify the desired line-wrap length for the `l' command
  --posix
                 disable all GNU extensions.
  -r, --regexp-extended
                 use extended regular expressions in the script.
  -s, --separate
                 consider files as separate rather than as a single continuous
                 long stream.
  -u, --unbuffered
                 load minimal amounts of data from the input files and flush
                 the output buffers more often
      --help     display this help and exit
      --version  output version information and exit

If no -e, --expression, -f, or --file option is given, then the first
non-option argument is taken as the sed script to interpret.  All
remaining arguments are names of input files; if no input files are
specified, then the standard input is read.

GNU sed home page: <http://www.gnu.org/software/sed/>.
General help using GNU software: <http://www.gnu.org/gethelp/>.
sh: 12: /opt/ros/hydro/share/rosbash/roszsh: Syntax error: "}" unexpected
%{%}%M:%1~%(!.#.$)%{%} exit
vm-ros:devel$ . "/opt/ros/hydro/share/rosbash/roszsh" 
vm-ros:devel$ 
vm-ros:devel$ 
```

**Comment by dirk-thomas** [comment]
I think I do understand what happens for you:

Basically after source `/opt/ros/hydro/setup.zsh` the variable `CATKIN_SHELL` is set to `zsh` even if it is not visible in the output of e.g. `env`. Then you invoke the Python script to build the new workspace. Python then invokes the generated `env.sh` file. Usually the not exported variable `CATKIN_SHELL` should not be available in the scope of the `env.sh`. But obviously it is for you.

@prfraanje Can you please test it with the catkin version from the `reset_shell` branch (https://github.com/ros/catkin/commits/reset_shell) - just check it out into your workspace - it will overlay the installed version. It has this fix: https://github.com/ros/catkin/commit/ec4efb6fe6971a9d34cfd3c325dbdd756352a662

@wjwwood I am not sure though why you are not having the same problem. May be the environment handle is a bit different?

**Comment by prfraanje** [comment]
@dirk-thomas, thank you for the fix, that made the job!

**Comment by prfraanje** [comment]
Unfortunately solution didn't work. I didn't tested well enough. I successfully compiled catkin-reset_shell in the bash-shell. Then I changed shell to zsh, and did again a catkin_make, but this make didn't need to do anything. After making creating a new package I still received the same error in zsh.

But, we came a step further. Now, I only needed to change the file `./src/catkin-reset_shell/cmake/templates/env.sh.in`. In this file, I changed the first line from `#!/usr/bin/env sh` into `#!/usr/bin/env zsh`. This fully did the job in zsh. I realize this is an ad hoc fix for zsh, but won't work in bash. @dirk-thomas, do you have an idea to get it completely fixed?

**Comment by prfraanje** [comment]
Related question: in my catkin_make output, I get many messages with `/home/rufus/catkin_ws/build/catkin_generated/setup_cached.sh:12: read-only variable: _`. Do I need to worry about these?

**Comment by dirk-thomas** [comment]
Changing the shebang line is not an option. Since I can't reproduce your issue it is difficult to give further advice. Have you checked the content of the generated env.sh file? With the patch it should never income any zsh specific code.

I have never seen the read-only messages. When are they shown after invoking what command? Something seems to be different on your system? Can you please describe what system you use and any custom non-standard setting etc.?

**Comment by prfraanje** [comment]
Starting again from a clean `~/.zshrc` config file and gradually building it up, I found out that my problem was due to the zsh-option `setopt all_export`. I think I copied this option once, but searching around I see now, that it is not recommended. Just export the zsh-variables with `export` explicitly. I also can compile with the original `env.sh` file and don't get the read-only messages anymore. Thanks for your advise!

**Comment by prfraanje** [comment]
Starting again from a clean `~/.zshrc` config file and gradually building it up, I found out that my problem was due to the zsh-option `setopt all_export`. I think I copied this option once, but searching around I see now, that it is not recommended. Just export the zsh-variables with `export` explicitly. I also can compile with the original `env.sh` file and don't get the read-only messages anymore. Thanks for your advise!

**Comment by prfraanje** [comment]
Also the old catkin_make, that is located in `/opt/ros/hydro`, works fine without using `setopt all_export` in `~/.zshrc`. This maybe explains why @wjwwood couldn't reproduce the error.

**Comment by dirk-thomas** [comment]
Thanks for digging deeper and commenting about the `setopt all_export` option. With that I was able to reproduce the problem directly.

I found a second related code block which needed the same fix: cf32d594b9e541363b4141b099b4920b437a0ab2 Even if the option is not recommended it should work fine now.
